trieste/space.py

```diff
@@ -212,15 +211,49 @@ def sample(self, num_samples: int, seed: Optional[int] = None) -> TensorType:
         :return: ``num_samples`` i.i.d. random points, sampled uniformly from this search space.
         """
 
+    def contains(self, value: TensorType) -> TensorType:
+        """Method for checking containment. Supports broadcasting.
+
+        :param value: A point or points to check for membership of this :class:`SearchSpace`.
+        :return: A boolean array showing membership for each point in value.
+        :raise ValueError (or tf.errors.InvalidArgumentError): If ``value`` has a different
+            dimensionality points from this :class:`SearchSpace`.
+        """
+        tf.debugging.assert_equal(
+            tf.rank(value) > 0 and tf.shape(value)[-1] == self.dimension,
+            True,
+            message=f"""
+                Dimensionality mismatch: space is {self.dimension}, value is {tf.shape(value)[-1]}
+                """,
+        )
+        return self._contains(value)
+
     @abstractmethod
-    def __contains__(self, value: TensorType) -> bool | TensorType:
+    def _contains(self, value: TensorType) -> TensorType:
+        """Space-specific implementation of containment. Can assume valid input shape.
+
+        :param value: A point or points to check for membership of this :class:`SearchSpace`.
+        :return: A boolean array showing membership for each point in value.
         """
-        :param value: A point to check for membership of this :class:`SearchSpace`.
-        :return: `True` if ``value`` is a member of this search space, else `False`. May return a
-            scalar boolean `TensorType` instead of the `bool` itself.
+
+    def __contains__(self, value: TensorType) -> bool:
+        """Method called by `in` operator. Doesn't support broadcasting as Python insists
+        on converting the result to a boolean.
+
+        :param value: A single point to check for membership of this :class:`SearchSpace`.
+        :return: `True` if ``value`` is a member of this search space, else `False`.
         :raise ValueError (or tf.errors.InvalidArgumentError): If ``value`` has a different
             dimensionality from this :class:`SearchSpace`.
         """
+        tf.debugging.assert_equal(
+            tf.rank(value) == 1,
+            True,
+            message=f"""
+                Rank mismatch: expected 1, got {tf.rank(value)}. To broadcast containment use
+                `space.contains(value)` method rather than `value in space`.
+                """,
+        )
+        return self.contains(value)
```

Comment:
this confused me? But I guess its for a good reason

Reply:
The problem is that the Python expression `A in B` automatically converts the result of calling `__contains__` to a boolean, which means it can't support broadcasting. That's why I've added a separate `contains` method that we can call to get broadcasting if we want to, while the `in` method will still give a useful error if you pass in multiple points.